pyqtorch/api.py

```diff
@@ -96,12 +99,98 @@ def sample(
+
+    Args:
+        circuit (QuantumCircuit): Quantum circuit :math:`U(\\theta)`.
+        state (Tensor): Input state :math:`\\ket\\rangle`.
+        observable (Observable): Observable O.
+        values (dict[str, Tensor], optional): Parameter values for the observable if any.
+        embedding (Embedding | None, optional): An optional instance of `Embedding`.
+
+    Returns:
+        Tensor: Expectation value.
+    """
+    state = run(circuit, state, values, embedding=embedding)
+    return inner_prod(state, run(observable, state, values, embedding=embedding)).real
+
+
+def sampled_expectation(
+    circuit: QuantumCircuit,
+    state: Tensor,
+    observable: Observable,
+    values: dict[str, Tensor] = dict(),
+    embedding: Embedding | None = None,
+    n_shots: int = 1,
+) -> Tensor:
+    """Expectation value approximated via sampling.
+
+    Given an input state :math:`\\ket\\rangle`,
+    the expectation analytical value with :math:`O` is defined as
+    :math:`\\langle\\bra|O\\ket\\rangle`
+
+    Args:
+        state (Tensor): Input state :math:`\\ket\\rangle`.
+        observable (Observable): Observable O.
+        values (dict[str, Tensor], optional): Parameter values for the observable if any.
+        embedding (Embedding | None, optional): An optional instance of `Embedding`.
+        n_shots: (int, optional): Number of samples to compute expectation on.
+
+    Returns:
+        Tensor: Expectation value.
+    """
+    state = run(circuit, state, values, embedding=embedding)
+    n_qubits = circuit.n_qubits
+
+    # batchsize needs to be first dim for eigh
+    eigvals, eigvecs = torch.linalg.eigh(
+        observable.tensor(n_qubits=n_qubits, values=values).permute((2, 0, 1))
+    )
+    eigvals = eigvals.squeeze()
+    eigvec_state_prod = apply_operator(
+        state,
+        eigvecs.T.conj(),
+        tuple(range(n_qubits)),
+        n_qubits=circuit.n_qubits,
+    )
+    eigvec_state_prod = torch.flatten(eigvec_state_prod, start_dim=0, end_dim=-2).t()
+    probs = torch.abs(torch.pow(eigvec_state_prod, 2))
+    batch_sample_multinomial = torch.func.vmap(
+        lambda p: sample_multinomial(
+            p, n_qubits, n_shots, return_counter=False, minlength=probs.shape[-1]
+        ),
+        randomness="different",
+    )
+    batch_samples = batch_sample_multinomial(probs)
+    normalized_samples = torch.div(
+        batch_samples, torch.tensor(n_shots, dtype=probs.dtype)
+    )
+    normalized_samples.requires_grad = True
+    expectations = torch.einsum(
+        "i,ji ->j", eigvals.to(dtype=normalized_samples.dtype), normalized_samples  # type: ignore[union-attr]
+    )
+    return expectations
+
+
 def expectation(
     circuit: QuantumCircuit,
     state: Tensor = None,
     values: dict[str, Tensor] = dict(),
     observable: Observable = None,  # type: ignore[assignment]
     diff_mode: DiffMode = DiffMode.AD,
+    options: dict[str, int] = dict(),
```

Comment:
Could it contain any other option than the number of shots ?

Reply:
@Roland-djee At the moment any extra information will be unused but I left it as it is to be flexible later, say we can pass a post-processing method or a custom expectation method. Would you suggest we just use for now `n_shots` as an argument of `expectation` instead of the `options` dict, and later adapt for future options?